### PR TITLE
Plugin manager system

### PR DESCRIPTION
**Pull request by danim1130**
The goal of this PR is to implement a plugin manager system, where distinct, separate functionalities of the Editor can be outsourced to smaller, modular code bases. This should reduce the complexities of the main project while providing a platform for outside developers to add or modify functionalities in the Editor without going through the main repository.

The system is still in development and heavily experimental.

In it's current form, the system should be capable to download and run the [Active Window WIP](https://github.com/intechstudio/plugin-active-win-wip) plugin. The functionality of this plugin should match the previous implementation.

- [x] After starting the application, a plugin list window should be shown on the preferences page, between the Profile Cloud URL and Profile window, with the "Active Window" plugin showing besides a download button.
- [x] Clicking the download button should start the downloading and extracting process. After a few seconds, the download should complete, and a check mark should appear besides the plugin title.
- [x] Pressing the uninstall button should return to the original state.
- [x] Pressing the check mark should enable the plugin.
    - [x] A new preferences window should appear under the plugin list window.
    - [x] This should be used to configure the plugin while also showing the current active window.
    - [x] The trigger application filter must exactly match the Active Window value
    - [x] The plugin should work just like the previous active window implementation
- [x] Disabling the plugin should stop the plugin functionality.

# Known errors

The plugin fails silently on MacOS if the necessary permissions (Accessibility and Screen Recording) are not given.

### CODE REVIEW

**Comment by narayb** [comment]
Everything works as described.

After Uninstall set values for active window titles remain in memory. If reinstalled within the same runtime, these titles remain visible. Is this intended?

File: package.json [hunk 1]
```diff
@@ -16,7 +16,7 @@
     "chokidar": "^3.4.3",
     "cross-env": "^7.0.2",
     "dotenv": "^14.2.0",
-    "electron": "^18.3.15",
+    "electron": "^22.0.0",
```
Comment (kkerti) [inline package.json:19]: sketchy move!

File: src/electron/plugin/pluginManager.ts [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+  const plugin = currentlyLoadedPlugins[pluginName];
+  await plugin.executeAction(actionId, payload);
+}
+
+async function notifyListener() {
+  const plugins = await getAvailablePlugins();
+  messagePort.postMessage({ type: "plugins", plugins });
+}
+
+async function getAvailablePlugins() {
```
Comment (kkerti) [inline src/electron/plugin/pluginManager.ts]: Kicsit szedd jobban szét a getAvailablePlugins-t üzleti logika jobban érződjön.

File: src/renderer/App.svelte [hunk 3]
```diff
@@ -109,6 +110,47 @@
     profileLinkStore.set({ id: value });
   });
 
+  window.onmessage = (event) => {
```
Comment (kkerti) [inline src/renderer/App.svelte:110]: Dobj bele egy-egy commentet, ha tovább bővül a funkcionalitás, akkor szedjük jobban szét.

File: src/renderer/main/panels/preferences/Preferences.svelte [hunk 4]
```diff
@@ -493,75 +543,46 @@
...
       <input
-        type="text"
-        placeholder="Page 3 trigger application"
         class="bg-primary my-1"
-        bind:value={$appSettings.persistant.pageActivatorCriteria_3}
-      />
+        type="checkbox"
+        checked={plugin.status === "Enabled"}
+        style="visibility:{plugin.status === "Downloaded" || plugin.status === "Enabled" ? "visible" : "hidden"}"
+        on:change={async e => {
```
Comment (kkerti) [inline src/renderer/main/panels/preferences/Preferences.svelte]: legyen ez egy onChangeHandler vagy hasonló

File: src/renderer/main/panels/preferences/Preferences.svelte [hunk 5]
```diff
@@ -493,75 +543,46 @@
...
+        class="flex items-center justify-center rounded my-2 focus:outline-none border-2 border-select bg-select hover:bg-select-saturate-10 hover:border-select-saturate-10 text-white px-2 py-0.5 mr-2"
+        on:click={window.pluginManagerPort.postMessage({type: "download-plugin", id : plugin.id})}
+        disabled={plugin.status == "Downloading"}
+        >
+        Download
+        </button>  
+      {:else}
+      <button 
+        class="flex items-center justify-center rounded my-2 focus:outline-none border-2 border-select bg-select hover:bg-select-saturate-10 hover:border-select-saturate-10 text-white px-2 py-0.5 mr-2"
+        on:click={window.pluginManagerPort.postMessage({type: "uninstall-plugin", id : plugin.id})}
```
Comment (kkerti) [inline src/renderer/main/panels/preferences/Preferences.svelte]: itt is handlerbe!

File: src/renderer/main/panels/preferences/Preferences.svelte [hunk 6]
```diff
@@ -493,75 +543,46 @@
...
+            window.pluginManagerPort.postMessage({type: "unload-plugin", id : plugin.id})
+          }
+        }}     
+    />
+    <div class="mx-1">{plugin.name}</div>
+    <div class="mx-1">
+      {#if plugin.status == "Downloading" || plugin.status == "Uninstalled" || plugin.status == "MarkedForDeletion"}
+      <button 
+        class="flex items-center justify-center rounded my-2 focus:outline-none border-2 border-select bg-select hover:bg-select-saturate-10 hover:border-select-saturate-10 text-white px-2 py-0.5 mr-2"
+        on:click={window.pluginManagerPort.postMessage({type: "download-plugin", id : plugin.id})}
```
Comment (kkerti) [inline src/renderer/main/panels/preferences/Preferences.svelte]: itt is handlerbe!

File: src/renderer/main/panels/preferences/Preferences.svelte [hunk 7]
```diff
@@ -493,75 +543,46 @@
...
+        >
+        Uninstall
+        </button>
+      {/if}
+    </div>
     </div>
+    {/each}
   </div>
 
+  <div id="pluginList" class="p-4 bg-secondary rounded-lg flex flex-col mb-4"/>
```
Comment (kkerti) [inline src/renderer/main/panels/preferences/Preferences.svelte]: `bind:this={VARIABLE}` https://svelte.dev/docs/element-directives#bind-this

**Review by kkerti — COMMENTED** [review]
Rendben.

**Comment by kkerti** [comment]
@danim1130 kérlek majd prettier-t futtass rajta és merge conflict legyen feloldva stable-el.

prettier: `prettier --write .`

File: package.json (hunk 1 above)
Comment (danim1130) [reply package.json:19]: Sketchy move for sketchy functionality

File: src/renderer/main/panels/preferences/Preferences.svelte (hunk 6 above)
Comment (kkerti) [reply src/renderer/main/panels/preferences/Preferences.svelte]:
```
on:click={(evt)=>{pluginManagerHandler("type",id)}}
```

**Review by kkerti — COMMENTED** [review]
ok!

**Pull request merged**